### ISSUE TITLE
Include resolve type in API documentation for Promise return types

### ISSUE THREAD

**Issue by fastfedora** [emberjs/ember.js#15114]
The API documentation for functions that use promises list `Promise` as the return type for the function (e.g. documentation for [DS.Store.findAll](https://emberjs.com/api/data/classes/DS.Store.html#method_findAll)). While _technically_ correct, this is nearly useless from a documentation perspective, since it doesn't tell you what type the promise resolves to. 

This becomes more noticeable when using the async/await syntax, where the promise gets hidden behind the scenes. For example:

```
let records = await store.findAll('users');
```

There's nothing in the documentation to indicate what the type of `records` is.

To improve the usability of the documentation, I propose including the type of the resolved value of the Promise in the `@return` tag so it appears in the generated API docs. 

One syntax would be to use the same syntax as generics, so to indicate that the promise resolves to a DS.Model object, the JSDocs would be:

 ```
/**
  * @return {Promise<DS.Model>} a promise which returns a single instance of the model`
  */
```

Which would then allow the API docs to show `Promise<DS.Model>` as the return type.

**Comment by pixelhandler** [comment]
@fastfedora would you consider taking a stab at a PR with the doc block changes. Looks like that is in the emberjs/data repo, see https://github.com/emberjs/data/blob/v2.12.0/addon/-private/system/store.js#L1556

**Comment by rwjblue** [comment]
> `@return {Promise<DS.Model>}`

Yep, I think this is 👍.  Would love to see PR's fixing things....

**Comment by pixelhandler** [comment]
@locks do you have an opinion on whether this type of documentation for the Promise's return value is included, e.g. `@return {Promise<DS.Model>}` vs `@return {Promise}` ? As @fastfedora mentions, using `await` does bring up the need.

**Comment by fastfedora** [comment]
@pixelhandler I've only been working with Ember less than a month, so I'm not the best person to go through the source code and update all the documentation, particularly since it requires unraveling calls to internal functions that are themselves not documented with resolve types.

For the `DS.Store.findAll` method, I did find that the regular documentation for [Models: Finding Records](https://guides.emberjs.com/v2.12.0/models/finding-records/) states "store.findAll() returns a DS.PromiseArray that fulfills to a DS.RecordArray". Thus the return type for this function should be `DS.PromiseArray<DS.RecordArray>`.

**Comment by pixelhandler** [comment]
@fastfedora I'll see if I can rally someone to work on this tomorrow.

**Comment by skaterdav85** [comment]
I'd be happy to work on this if someone isn't already.

**Comment by pixelhandler** [comment]
@skaterdav85 thanks for working on this, I tried to assign to you but didn't work for some reason, so assigned to myself, let me know If you need any help

**Comment by skaterdav85** [comment]
cool, thanks @pixelhandler!

**Comment by pixelhandler** [comment]
@fastfedora if you have any questions about Ember Data, @skaterdav85 wrote a great book on the subject :)

**Comment by rwjblue** [comment]
> I tried to assign to you but didn't work for some reason

yeah, GitHub only allows you to assign members of the org I think

**Comment by Serabe** [comment]
Fixed in #15316. Thank you!
